core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java

```diff
@@ -297,61 +484,38 @@ private void migrateFromV11() {
-               FileEntry newFe = new FileEntry(newFilePos, oldFe.size, newKey.limit(), newValue.limit(), newMetaSize,
-                     oldFe.internalMetadataLen, oldFe.expiryTime);
-
-               // write the FileEntry to new file
-               buf = allocate(buf, KEY_POS_LATEST);
-               newFe.writeToBuf(buf);
-               buf.flip();
-               newChannel.write(buf, newFilePos);
-               newFilePos += KEY_POS_LATEST; //size written
-               entries.put(key, newFe);
-
-               //write the updated content to the new file
-               newFilePos += newChannel.write(newKey, newFilePos);
-               newFilePos += newChannel.write(newValue, newFilePos);
-               if (newMeta != null) {
-                  newFilePos += newChannel.write(newMeta, newFilePos);
-               }
-               newChannel.write(newKey, newFilePos);
-               newFilePos += newFe.keyLen;
-               newChannel.write(newValue, newFilePos);
-               newFilePos += newFe.dataLen;
-               if (newMeta != null) {
-                  newChannel.write(newMeta, newFilePos);
-                  newFilePos += newFe.metadataLen;
+            long created = -1;
+            long lastUsed = -1;
+            bufRef.set(buf);
+            K key = unmarshallObject(bufRef, offset, oldFe.keyLen, wrapperMissing);
+            V value = unmarshallObject(bufRef, offset, oldFe.dataLen, wrapperMissing);
+            Metadata metadata = null;
+            if (oldFe.metadataLen > 0) {
+               metadata = unmarshallObject(bufRef, offset, oldFe.metadataLen - TIMESTAMP_BYTES, wrapperMissing);
+
+               if (oldFe.expiryTime > 0) {
+                  buf = bufRef.get();
+                  buf = readChannelUpdateOffset(buf, offset, TIMESTAMP_BYTES);
+                  buf.flip();
+                  created = buf.getLong();
+                  lastUsed = buf.getLong();
+                  bufRef.set(buf);
                }
-
-               remainingLength = (newFe.expiryTime > 0 ? TIMESTAMP_BYTES : 0) + oldFe.internalMetadataLen + 8;
-            } else {
-               // Simply use the old FileEntry as only the magic bytes are updated
-               entries.put(key, oldFe);
-               remainingLength = oldFe.size - oldFe.keyLen - KEY_POS_11_0;
             }
 
-            // Read and write the remaining data from old file
-            buf = readChannelUpdateOffset(buf, offset, remainingLength);
-            newFilePos += newChannel.write(buf, newFilePos);
+            PrivateMetadata internalMeta = null;
+            if (oldFe.internalMetadataLen > 0) {
+               internalMeta = unmarshallObject(bufRef, offset, oldFe.internalMetadataLen, wrapperMissing);
+               buf = bufRef.get();
+            }
+            MarshallableEntry<? extends K, ? extends V> me = (MarshallableEntry<? extends K, ? extends V>) ctx.getMarshallableEntryFactory()
```

Comment:
We may want to add the same code as https://github.com/infinispan/infinispan/pull/9409/files#diff-3724b4c53715346242997a7b54b4c3a863ae350dd5f3911a9fb2ce0f951db398R421 so we don't write an expired entry to the file or index.

This shouldn't affect correctness though as the entry will just be ignored later when it is found to be expired, since we have the correct created and lastUsed times.